extensions/ipynb/src/notebookImagePaste.ts

```diff
@@ -0,0 +1,145 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+
+class CopyPasteEditProvider implements vscode.DocumentPasteEditProvider {
+
+	async provideDocumentPasteEdits(
+		_document: vscode.TextDocument,
+		_ranges: readonly vscode.Range[],
+		dataTransfer: vscode.DataTransfer,
+		_token: vscode.CancellationToken
+	): Promise<vscode.DocumentPasteEdit | undefined> {
+
+		// get foundational notebook data
+		const activeNotebook = vscode.window.activeNotebookEditor?.notebook;
+		const activeSelection = vscode.window.activeNotebookEditor?.selection;
+		const currentCell = activeNotebook?.getCells(activeSelection);
+		if (!currentCell || currentCell?.length !== 1) {
+			return undefined;
+		}
+		const initialAttachments = currentCell[0].metadata?.custom?.attachments;
+		const initialize = !initialAttachments ? true : false;
+
+		// get b64 data from paste
+		const dataItem = dataTransfer.get('image/png');
+		if (!dataItem) {
+			return undefined;
+		}
+		const fileDataAsUint8 = await dataItem.asFile()?.data();
+		if (!fileDataAsUint8) {
+			return undefined;
+		}
+		const b64string = encodeBase64(fileDataAsUint8);
+
+		// get filename data from paste
+		let pasteFilename = dataItem.asFile()?.name;
+		const filePrefix = pasteFilename?.split('.')[0];
+		const fileSuffix = pasteFilename?.split('.')[1];
+		if (!filePrefix || !fileSuffix) {
+			return undefined;
+		}
+
+		// create updated metadata for cell (prep for WorkspaceEdit)
+		if (!initialize) {
+			let appendValue = 2;
+			let attachmentFilename: keyof typeof initialAttachments;
+			const sortedFilenames = sortFilenames(initialAttachments);
+			for (attachmentFilename of sortedFilenames) {
+				if (attachmentFilename === pasteFilename) {
+					const objEntries = Object.entries(initialAttachments[pasteFilename]);
+					if (objEntries.length) { // check that mime:b64 are present
+						const [, attachmentb64] = objEntries[0];
+						if (attachmentb64 !== b64string) {	// append a "-#" here. same name, diff data. this matches jupyter behavior
+							pasteFilename = filePrefix.concat(`-${appendValue++}.`) + fileSuffix;
+						}
+					}
+				}
+			}
+		}
+		if (!pasteFilename) {
+			return undefined;
+		}
+		const updatedMetadata = { 'custom': { 'attachments': { [pasteFilename]: { 'image/png': b64string } } } };
+		if (!initialize) {
+			let attachmentFilename: keyof typeof initialAttachments;
+			for (attachmentFilename in initialAttachments) {
+				updatedMetadata.custom.attachments[attachmentFilename] = initialAttachments[attachmentFilename];
+			}
+		}
+
+		// create WorkspaceEdit and apply to active notebook
+		let newCell;
+		if (updatedMetadata) {
+			newCell = vscode.NotebookEdit.updateCellMetadata(currentCell[0].index, updatedMetadata);
+		}
+		const edit = new vscode.WorkspaceEdit();
+		if (activeNotebook?.uri && newCell) {
+			edit.set(activeNotebook.uri, [newCell]);
+		}
+		vscode.workspace.applyEdit(edit);
```

Comment:
The editor should be returned as part of the result instead of applied here

Reply:
👍 for merging this into the returned edit, to ensure Undo/Redo work fine. 

@mjbvz However, it seems the document paste provider will run `reviveWorkspaceEditDto` against the edits, which doesn't support notebook cell edits. The bulk edit service is now using `reviveWorkspaceEditDto2`, we might want to use that

Reply:
This sounds reasonable to me. Should we include that change in this PR too?